subsys/emul/i2c/emul_sbs_gauge.c

```diff
@@ -0,0 +1,188 @@
+	case SBS_GAUGE_CMD_VOLTAGE:
+	case SBS_GAUGE_CMD_AVG_CURRENT:
+	case SBS_GAUGE_CMD_TEMP:
+	case SBS_GAUGE_CMD_ASOC:
+	case SBS_GAUGE_CMD_FULL_CAPACITY:
+	case SBS_GAUGE_CMD_REM_CAPACITY:
+	case SBS_GAUGE_CMD_NOM_CAPACITY:
+	case SBS_GAUGE_CMD_AVG_TIME2EMPTY:
+	case SBS_GAUGE_CMD_AVG_TIME2FULL:
+	case SBS_GAUGE_CMD_CYCLE_COUNT:
+	case SBS_GAUGE_CMD_DESIGN_VOLTAGE:
+		val = 1;
+		break;
+	default:
+		LOG_INF("Unknown read %x", regn);
+	}
+	LOG_INF("       = %x", val);
+
+	return val;
+}
+
+static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg *msgs,
+				       int num_msgs, int addr)
+{
+	struct sbs_gauge_emul_data *data;
+	unsigned int val;
+
+	data = target->data;
+
+	__ASSERT_NO_MSG(msgs && num_msgs);
+
+	i2c_dump_msgs("emul", msgs, num_msgs, addr);
+	switch (num_msgs) {
+	case 2:
+		if (msgs->flags & I2C_MSG_READ) {
+			LOG_ERR("Unexpected read");
+			return -EIO;
+		}
+		if (msgs->len != 1) {
+			LOG_ERR("Unexpected msg0 length %d", msgs->len);
+			return -EIO;
+		}
+		data->cur_reg = msgs->buf[0];
+
+		/* Now process the 'read' part of the message */
+		msgs++;
+		if (msgs->flags & I2C_MSG_READ) {
+			switch (msgs->len - 1) {
+			case 1:
+				val = reg_read(target, data->cur_reg);
+				msgs->buf[0] = val;
+				break;
+			default:
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+				return -EIO;
+			}
+		} else {
+			if (msgs->len != 1) {
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+			}
+			reg_write(target, data->cur_reg, msgs->buf[0]);
+		}
+		break;
+	default:
+		LOG_ERR("Invalid number of messages: %d", num_msgs);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/* Device instantiation below */
+
+static struct i2c_emul_api sbs_gauge_emul_api_i2c = {
+	.transfer = sbs_gauge_emul_transfer_i2c,
+};
+
+/**
+ * Set up a new SBS_GAUGE emulator (I2C)
+ *
+ * This should be called for each SBS_GAUGE device that needs to be emulated. It
+ * registers it with the i2c emulation controller.
+ *
+ * @param emul Emulation information
+ * @param parent Device to emulate (must use sbs_gauge driver)
+ * @return 0 indicating success (always)
+ */
+static int emul_sbs_sbs_gauge_init_i2c(const struct emul *target, const struct device *parent)
+{
+	ARG_UNUSED(target);
+	ARG_UNUSED(parent);
+
+	return 0;
+}
+
+#define SBS_GAUGE_EMUL_DATA(n)                                                                     \
+	static uint8_t sbs_gauge_emul_reg_##n[2];                                                  \
+	static struct sbs_gauge_emul_data sbs_gauge_emul_data_##n;
+
+#define SBS_GAUGE_EMUL_DEFINE(n, type)                                                             \
+	EMUL_DEFINE(emul_sbs_sbs_gauge_init_##type, DT_DRV_INST(n), &sbs_gauge_emul_cfg_##n,       \
+		    &sbs_gauge_emul_data_##n, &sbs_gauge_emul_api_i2c)
+
+#define SBS_GAUGE_EMUL_I2C(n)                                                                      \
+	SBS_GAUGE_EMUL_DATA(n)                                                                     \
```

Comment:
IMO, the amount of indirection here is unnecessary and makes this a little bit harder to follow than otherwise. I take it that, at some point, you'll want to parameterize this instantiation based on whether the gauge is on the I2C or SPI buses.

I'd suggest just not worrying about that for now and putting the entire contents of these instantiation macros in `SBS_GAUGE_EMUL`. But if you do want to worry about it, I'd try to get rid of `SBS_GAUGE_EMUL_DEFINE` and `SBS_GAUGE_EMUL_DATA` and just write `SBS_GAUGE_EMUL` so that it has what it needs to appropriately parameterize `EMUL_DEFINE`.

Reply:
Done

Reply:
I still see `SBS_GAUGE_EMUL_DEFINE` and `SBS_GAUGE_EMUL_DATA` as separate macros.